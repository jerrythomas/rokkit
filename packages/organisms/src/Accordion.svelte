<script>
	import { createEmitter } from '@rokkit/core'
	// import { createEventDispatcher } from 'svelte'
	import { defaultFields } from '@rokkit/core'
	import { navigator } from '@rokkit/actions'
	import { Item, Summary } from '@rokkit/molecules'
	// import List from './List.svelte'
	import ListItems from './ListItems.svelte'

	// const dispatch = createEventDispatcher()

	/**
	 * @typedef {Object} Props
	 * @property {string} [class]
	 * @property {any} [items]
	 * @property {any} [fields]
	 * @property {any} [using]
	 * @property {boolean} [autoClose]
	 * @property {any} [value]
	 */

	/** @type {Props} */
	let {
		class: className = '',
		items = $bindable([]),
		fields = $bindable({}),
		using = $bindable({}),
		autoClose = false,
		value = $bindable(null),
		...events
	} = $props()
	let cursor = $state([])

	let emitter = $derived(createEmitter(events, ['collapse', 'change', 'expand', 'click']))
	let internalFields = $derived({ ...defaultFields, ...fields })
	let internalUsing = $derived({ default: Item, ...using })

	function handle(event) {
		value = event.detail.node
		cursor = event.detail.path
		if (['collapse', 'expand'].includes(event.type)) {
			if (autoClose) {
				items.map((x) => {
					if (x !== value && x[internalFields.isOpen]) {
						x[internalFields.isOpen] = false
					}
				})
			}
			items = items
		}
		emitter[event.type](event.detail)
		// dispatch(event.type, event.detail)
	}
</script>

<!-- svelte-ignore a11y_no_noninteractive_tabindex -->
<accordion
	class="flex w-full select-none flex-col {className}"
	tabindex="0"
	use:navigator={{
		items,
		fields: internalFields,
		enabled: true,
		indices: cursor
	}}
	onselect={handle}
	onmove={handle}
	onexpand={handle}
	oncollapse={handle}
>
	{#each items as item, index}
		{@const hasItems = item[internalFields.children] && item[internalFields.children].length > 0}
		{@const itemFields = { ...internalFields, ...(internalFields.fields ?? internalFields) }}

		<div
			id={'id-' + index}
			class="flex flex-col"
			class:is-expanded={item[internalFields.isOpen]}
			class:is-selected={item === value}
			data-path={index}
		>
			<Summary fields={internalFields} {using} bind:value={items[index]} />
			{#if hasItems && item[internalFields.isOpen]}
				<list class="flex w-full flex-shrink-0 select-none flex-col" role="listbox" tabindex="-1">
					<ListItems
						bind:items={item[internalFields.children]}
						bind:value
						fields={itemFields}
						{using}
						hierarchy={[index]}
						on:change
					/>
				</list>
				<!-- <List
					bind:items={item[fields.children]}
					bind:value
					fields={itemFields}
					{using}
					on:select
					hierarchy={[index]}
					tabindex="-1"
				/> -->
			{/if}
		</div>
	{/each}
</accordion>
