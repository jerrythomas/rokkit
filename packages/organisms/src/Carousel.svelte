<script>
	import { equals } from 'ramda'
	/**
	 * @typedef {Object} Props
	 * @property {string} [class]
	 * @property {any} [items]
	 * @property {any} value
	 */

	/** @type {Props} */
	let { class: className = '', items = [], value = $bindable() } = $props()
	let currentIndex = $state(0)

	function handleKey(e, index) {
		const prevIndex = currentIndex
		if (['Enter', ' '].includes(e.key)) {
			currentIndex = index
		} else {
			if (e.key === 'ArrowRight') {
				currentIndex = currentIndex === items.length - 1 ? 0 : currentIndex + 1
			} else if (e.key === 'ArrowLeft') {
				currentIndex = currentIndex === 0 ? items.length - 1 : currentIndex - 1
			}
		}
		if (currentIndex !== prevIndex) {
			e.preventDefault()
			e.stopPropagation()
		}
	}
	$effect(() => {
		value = items[currentIndex]
	})
</script>

{#if items.length === 0}
	<p>No items to display</p>
{:else}
	<carousel class="flex w-full flex-col gap-3 {className}">
		<slide class="flex flex-col">
			<img src={value.src} alt={value.alt} />
			<p>{value.text}</p>
		</slide>
		<dot-nav role="radiogroup">
			{#each items as item, index}
				<dot
					role="radio"
					aria-checked={equals(currentIndex, index)}
					aria-label={`Slide ${index + 1}`}
					onclick={() => (currentIndex = index)}
					onkeydown={(e) => handleKey(e, index)}
					tabindex={0}
				></dot>
			{/each}
		</dot-nav>
	</carousel>
{/if}
