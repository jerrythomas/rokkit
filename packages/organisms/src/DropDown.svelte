<script>
	import { defaultFields, defaultStateIcons, createEmitter } from '@rokkit/core'
	import { dismissable } from '@rokkit/actions'
	import { Icon, Slider } from '@rokkit/atoms'
	import { Item } from '@rokkit/molecules'
	import List from './List.svelte'

	/**
	 * @typedef {Object} Props
	 * @property {string} [class]
	 * @property {any} [options]
	 * @property {import('@rokkit/core').FieldMapping} [fields]
	 * @property {any} [using]
	 * @property {any} [value]
	 * @property {any} [title]
	 * @property {any} [icon]
	 * @property {boolean} [small]
	 */

	/** @type {Props} */
	let {
		class: className = '',
		options = [],
		fields = $bindable(defaultFields),
		using = $bindable({ default: Item }),
		value = $bindable(null),
		title = null,
		icon = null,
		small = false,
		...events
	} = $props()

	// $effect.pre(() => {
	// using = { default: Item, ...using }
	// fields = { ...defaultFields, ...fields }
	// })
	let internalUsing = $derived({ default: Item, ...using })
	let internalFields = $derived({ ...defaultFields, ...fields })

	let emitter = createEmitter(events, ['change'])
	let offsetTop = $state(0)
	let open = $state(false)
	let icons = defaultStateIcons.selector

	function handleSelect(data) {
		open = false
		emitter.change(data)
	}
</script>

<!-- svelte-ignore a11y_no_noninteractive_tabindex -->
<drop-down
	class="dropdown relative flex cursor-pointer select-none {className}"
	class:open
	aria-haspopup="true"
	aria-controls="menu"
>
	<button
		class="flex items-center"
		bind:clientHeight={offsetTop}
		tabindex="0"
		use:dismissable
		ondismiss={() => (open = false)}
		onfocus={() => (open = true)}
		onblur={() => (open = false)}
	>
		<span class="flex items-center">
			{#if icon !== null}
				<Icon name={icon} />
			{/if}
			{#if !small && title}
				<p class="flex w-full">{title}</p>
			{/if}
			{#if open}
				<Icon name={icons.opened} />
			{:else}
				<Icon name={icons.closed} />
			{/if}
		</span>
	</button>
	{#if open}
		<Slider top={offsetTop + 4}>
			<List
				items={options}
				fields={internalFields}
				using={internalUsing}
				bind:value
				onselect={handleSelect}
				tabindex="-1"
			/>
		</Slider>
	{/if}
</drop-down>
