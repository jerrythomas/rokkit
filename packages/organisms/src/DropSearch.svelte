<script>
	import { defaultFields } from '@rokkit/core'
	import List from './List.svelte'
	import { Slider } from '@rokkit/atoms'
	import { Item } from '@rokkit/molecules'

	export let items
	export let value = null
	/** @type {import('@rokkit/core').FieldMapping} */
	export let fields = defaultFields
	export let using = { default: Item }

	$: using = { default: Item, ...using }
	$: fields = { ...defaultFields, ...fields }

	let opened = false
	let search
	let searchBox

	// on search change filter list
	function handleSelect(event) {
		console.log('selected', event)
		value = event.detail
		search = event.detail[fields.text]
	}
</script>

<!-- svelte-ignore a11y-no-noninteractive-tabindex -->
<div
	class="dropdown relative h-12 w-full flex flex-col outline-none"
	tabindex={0}
>
	<div class="selected-item h-12 flex flex-shrink-0 items-center pl-4">
		<span class="flex flex-grow">
			<input
				type="text"
				class="w-full border-none bg-transparent p-0"
				bind:value={search}
				bind:this={searchBox}
				on:focus={() => {
					opened = true
					searchBox.select()
				}}
				on:blur={() => {
					opened = false
				}}
			/>
		</span>
		{#if opened}
			<icon class="selector-opened" />
		{:else}
			<icon class="selector-closed" />
		{/if}
	</div>

	{#if opened}
		<Slider>
			<List
				bind:items
				bind:value
				{fields}
				{using}
				on:select={handleSelect}
				on:change
			/>
		</Slider>
	{/if}
</div>
