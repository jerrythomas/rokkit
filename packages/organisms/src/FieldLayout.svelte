<script>
	import FieldLayout from './FieldLayout.svelte'
	import { getContext } from 'svelte'
	import { omit } from 'ramda'
	import InputField from './InputField.svelte'
	import { createEmitter } from '@rokkit/core'
	// const dispatch = createEventDispatcher()
	const registry = getContext('registry')

	let { value = $bindable({}), schema = {}, path = [], ...events } = $props()

	let emitter = $derived(createEmitter(events, ['change']))
	function handle() {
		emitter.change(value)
	}

	let wrapper = $derived(registry.wrappers[schema.wrapper] ?? registry.wrappers.default)
	let wrapperProps = $derived(omit(['wrapper', 'elements', 'key'], schema))
</script>

{#if !Array.isArray(schema.elements)}
	<error> Invalid schema. Expected schema to include an 'elements' array. </error>
{:else}
	{@const SvelteComponent_1 = wrapper}
	<SvelteComponent_1 {...wrapperProps}>
		{#each schema.elements as item}
			{@const elementPath = item.key ? [...path, item.key] : path}
			{@const props = { ...item.props, path: elementPath }}
			{@const nested = Array.isArray(item.elements) && item.elements.length > 0}
			{@const component = item.component
				? (registry.components[item.component] ?? registry.components.default)
				: null}

			{#if nested}
				{#if item.key}
					<FieldLayout {...props} schema={item} bind:value={value[item.key]} on:change={handle} />
				{:else}
					<FieldLayout {...props} schema={item} bind:value on:change={handle} />
				{/if}
			{:else if component}
				{@const SvelteComponent = component}
				<SvelteComponent {...item.props} value={item.key ? value[item.key] : null} />
			{:else}
				{@const name = elementPath.join('.')}
				<InputField {name} bind:value={value[item.key]} {...item.props} on:change={handle} />
			{/if}
		{/each}
	</SvelteComponent_1>
{/if}
