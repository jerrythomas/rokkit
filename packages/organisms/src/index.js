export * from './lib'
export { default as Accordion } from './Accordion.svelte'
export { default as DropDown } from './DropDown.svelte'
export { default as DropSearch } from './DropSearch.svelte'
export { default as DataEditor } from './DataEditor.svelte'
export { default as Form } from './Form.svelte'
export { default as List } from './List.svelte'
export { default as MultiSelect } from './MultiSelect.svelte'
export { default as NestedList } from './NestedList.svelte'
export { default as Select } from './Select.svelte'
export { default as Switch } from './Switch.svelte'
export { default as Tabs } from './Tabs.svelte'
export { default as Tree } from './Tree.svelte'
export { default as InputField } from './InputField.svelte'
export { default as ListEditor } from './ListEditor.svelte'
export { default as VirtualSelect } from './VirtualSelect.svelte'
export { default as VirtualList } from './VirtualList.svelte'
export { default as NestedEditor } from './NestedEditor.svelte'
export { default as Table } from './Table.svelte'
export { default as TreeTable } from './TreeTable.svelte'
