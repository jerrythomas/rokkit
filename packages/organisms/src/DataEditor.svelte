<script>
	import { run } from 'svelte/legacy'

	import { setContext } from 'svelte'
	import { writable } from 'svelte/store'
	import { componentTypes } from './types'
	import Wrapper from './wrappers/Wrapper.svelte'
	import { Item } from '@rokkit/molecules'
	import Tabs from './Tabs.svelte'
	import FieldLayout from './FieldLayout.svelte'
	import { createEventDispatcher } from 'svelte'

	const dispatch = createEventDispatcher()
	const registry = writable({})
	setContext('registry', registry)

	import { deriveSchemaFromValue, deriveLayoutFromValue, getSchemaWithLayout } from './lib'

	/**
	 * @typedef {Object} Props
	 * @property {any} value
	 * @property {any} [schema]
	 * @property {any} [layout]
	 * @property {any} [using]
	 */

	/** @type {Props} */
	let {
		value = $bindable(),
		schema = $bindable(null),
		layout = $bindable(null),
		using = {}
	} = $props()

	let schemaWithLayout = $derived(getSchemaWithLayout(schema, layout))

	function handle() {
		dispatch('change', value)
	}
	$effect(() => {
		registry.set({
			editors: { ...componentTypes, ...using?.editors },
			components: { default: Item, ...using?.components },
			wrappers: { default: Wrapper, ...using?.wrappers },
			navigators: { default: Tabs, ...using?.navigators }
		})
	})
	$effect(() => {
		if (!schema) schema = deriveSchemaFromValue(value)
	})
	$effect(() => {
		if (!layout) layout = deriveLayoutFromValue(value)
	})
</script>

<FieldLayout schema={schemaWithLayout} bind:value on:change={handle} />
