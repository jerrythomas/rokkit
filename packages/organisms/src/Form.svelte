<script>
	import { createBubbler } from 'svelte/legacy'

	const bubble = createBubbler()
	import DataEditor from './DataEditor.svelte'

	/**
	 * @typedef {Object} Props
	 * @property {any} value
	 * @property {any} [schema]
	 * @property {any} [layout]
	 * @property {any} [using]
	 * @property {import('svelte').Snippet} [children]
	 */

	/** @type {Props} */
	let { value = $bindable(), schema = null, layout = null, using = {}, children } = $props()
</script>

<form onsubmit={bubble('submit')}>
	<DataEditor bind:value {schema} {layout} {using} />
	<span>
		{#if children}
			{@render children()}
		{:else}
			<button type="submit">Submit</button>
		{/if}
	</span>
</form>
