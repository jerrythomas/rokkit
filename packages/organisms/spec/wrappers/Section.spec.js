import { describe, expect, it, beforeEach } from 'vitest'
import { cleanup, render } from '@testing-library/svelte'
import { tick } from 'svelte'
import { Section } from '../../src/wrappers'

describe('Section.svelte', () => {
	beforeEach(() => cleanup())

	it('should render using default props', () => {
		const { container } = render(Section)
		expect(container).toBeTruthy()
		expect(container).toMatchSnapshot()
	})

	it('should render using props', async () => {
		const { container, component } = render(Section, {
			title: 'Section Title',
			open: false
		})
		expect(container).toBeTruthy()
		expect(container).toMatchSnapshot()

		component.$set({ title: 'custom' })
		await tick()
		expect(container).toMatchSnapshot()
	})

	it('should render using specified type', async () => {
		const { container, component } = render(Section, { type: 'horizontal' })
		expect(container).toBeTruthy()
		expect(container).toMatchSnapshot()

		component.$set({ type: 'section' })
		await tick()
		expect(container).toMatchSnapshot()
	})

	it('should render using specified class', async () => {
		const { container, component } = render(Section, { class: 'custom-class' })
		expect(container).toBeTruthy()
		expect(container).toMatchSnapshot()

		component.$set({ class: 'custom-class-2' })
		await tick()
		expect(container).toMatchSnapshot()
	})
})
