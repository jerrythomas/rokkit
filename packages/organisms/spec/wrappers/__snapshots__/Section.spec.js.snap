// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Section.svelte > should render using default props 1`] = `
<body>
  <div>
    <details
      class=""
      open=""
    >
      <summary>
        Section
      </summary>
       
      <field-layout
        class="vertical"
      />
    </details>
  </div>
</body>
`;

exports[`Section.svelte > should render using props 1`] = `
<body>
  <div>
    <details
      class=""
    >
      <summary>
        Section Title
      </summary>
       
      <field-layout
        class="vertical"
      />
    </details>
  </div>
</body>
`;

exports[`Section.svelte > should render using props 2`] = `
<body>
  <div>
    <details
      class=""
    >
      <summary>
        custom
      </summary>
       
      <field-layout
        class="vertical"
      />
    </details>
  </div>
</body>
`;

exports[`Section.svelte > should render using specified class 1`] = `
<body>
  <div>
    <details
      class="custom-class"
      open=""
    >
      <summary>
        Section
      </summary>
       
      <field-layout
        class="vertical"
      />
    </details>
  </div>
</body>
`;

exports[`Section.svelte > should render using specified class 2`] = `
<body>
  <div>
    <details
      class="custom-class-2"
      open=""
    >
      <summary>
        Section
      </summary>
       
      <field-layout
        class="vertical"
      />
    </details>
  </div>
</body>
`;

exports[`Section.svelte > should render using specified type 1`] = `
<body>
  <div>
    <details
      class=""
      open=""
    >
      <summary>
        Section
      </summary>
       
      <field-layout
        class="horizontal"
      />
    </details>
  </div>
</body>
`;

exports[`Section.svelte > should render using specified type 2`] = `
<body>
  <div>
    <details
      class=""
      open=""
    >
      <summary>
        Section
      </summary>
       
      <field-layout
        class="section"
      />
    </details>
  </div>
</body>
`;
